src/groth16/verifier/batch.rs

```diff
@@ -167,4 +173,117 @@ where
             Err(VerificationError::InvalidProof)
         }
     }
+
+    /// Perform batch verification with a particular `VerifyingKey`, returning
+    /// `Ok(())` if all proofs were verified and `VerificationError` otherwise.
+    ///
+    /// This performs the bulk of internal arithmetic over the global rayon
+    /// threadpool.
+    #[cfg(feature = "multicore")]
+    #[allow(non_snake_case)]
+    pub fn verify_multicore(self, vk: &VerifyingKey<E>) -> Result<(), VerificationError> {
+        if self
+            .items
+            .iter()
+            .any(|Item { inputs, .. }| inputs.len() + 1 != vk.ic.len())
+        {
+            return Err(VerificationError::InvalidVerifyingKey);
+        }
+
+        struct Accumulator<E: MultiMillerLoop> {
+            gammas: Vec<E::Fr>,
+            delta: E::G1,
+            y: E::Fr,
+            ml_result: Option<E::Result>,
+        }
+
+        impl<E: MultiMillerLoop> Accumulator<E> {
+            fn new(ic_len: usize) -> Self {
+                Accumulator {
+                    gammas: vec![E::Fr::zero(); ic_len],
+                    delta: E::G1::identity(),
+                    y: E::Fr::zero(),
+                    ml_result: None,
+                }
+            }
+        }
+
+        let ic_len = vk.ic.len();
+
+        let acc = self
+            .items
+            // This chunk size was obtained heuristically.
+            .par_chunks(8)
```

Comment:
I determined this by running the [`connectblocksapling` benchmark for batched Sapling proof validation](https://github.com/zcash/zcash/pull/6048) on a Ryzen 9 5950X while adjusting this value. I got the following results:
| Chunk size | Average benchmark time (s) |
|---|---|
| 1   | 1.860 |
| 2   | 1.855 |
| 4   | 1.848 |
| **8**   | **1.802** |
| 12  | 1.815 |
| 16  | 1.856 |
| 32  | 1.844 |
| 64  | 1.904 |
| 80  | 1.929 |
| 96  | 1.970 |
| 112 | 1.972 |
| 128 | 2.005 |

Reply:
Is this likely to vary by CPU, and if so should that degree of parallelism be configurable by the caller?

Reply:
The degree of parallelism is configurable via `RAYON_NUM_THREADS` as this uses the global Rayon threadpool. This parameter just controls the relationship between `multi_miller_loop` (which has more benefit with larger chunks) vs work stealing (which has more benefit with smaller chunks). There might be an effect by CPU, but it's not one that can usefully be configured I suspect, and as you can see above the effect is relatively small (compared to the number of threads, as going from 1 to 32 gave a much larger magnitude of change to the benchmark).